websockets/websockets.go

```diff
@@ -539,6 +420,109 @@ func (w *webSocket) queueMessage(msg *message) {
 	})
 }
 
+func (w *webSocket) readPump(wg *sync.WaitGroup) {
+	defer wg.Done()
+	for {
+		messageType, data, err := w.conn.ReadMessage()
+		if err != nil {
+			if !websocket.IsUnexpectedCloseError(
+				err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
+				w.tq.Queue(func() error {
+					return w.connectionClosedWithError(nil)
+				})
+				return
+			}
+			w.tq.Queue(func() error {
+				_ = w.conn.Close() // TODO fix this
+				return w.connectionClosedWithError(err)
+			})
+			return
+		}
+
+		w.queueMessage(&message{
+			mtype: messageType,
+			data:  data,
+			t:     time.Now(),
+		})
```

Comment:
How about switching the condition logic, which should decrease the nesting and improve the readability? 

Also, should we do `w.conn.Close()` on the event loop (but this is probably not only this place)? :thinking:

```suggestion
		if err == nil {
			w.queueMessage(&message{
				mtype: messageType,
				data:  data,
				t:     time.Now(),
			})
			
			continue
		}
		
		if !websocket.IsUnexpectedCloseError(
			err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
			// maybe still log it with debug level?
			err = nil
		}
		
		if err != nil {
			w.tq.Queue(func() error {
				_ = w.conn.Close() // TODO fix this
				return nil
			})    
		}
		
		w.tq.Queue(func() error {
			return w.connectionClosedWithError(err)
		})
		
		return
		
```